### ISSUE TITLE
Ash Compiles before Plug breaking Ash JSON API

### ISSUE THREAD

**Issue by rowlandau** [ash-project/ash#449]
**Describe the bug**
While compiling AshHq,
Module `Ash.PlugHelpers` does not load, preventing the compilation of `:ash_json_api`.

```
== Compilation error in file lib/ash_json_api/request.ex ==
** (CompileError) lib/ash_json_api/request.ex:20: module Ash.PlugHelpers is not loaded and could not be found. This may be happening because the module you are trying to load directly or indirectly depends on the current module

could not compile dependency :ash_json_api, "mix compile" failed. Errors may have been logged above. You can recompile this dependency with "mix deps.compile ash_json_api", update it with "mix deps.update ash_json_api" or clean it with "mix deps.clean ash_json_api"
```

**To Reproduce**
```
git clone https://github.com/ash-project/ash_hq
cd ash_hq/
mix deps.get
mix compile
```

**Expected behavior**
AshHq compiles without issue.

**Runtime**
 - Elixir version: 1.14.0 (compiled with Erlang/OTP 23)
 - Erlang version: Erlang/OTP 24, erts-12.3.2
 - OS: macOS 12.3
 - Ash version: main - commit 5d434448efc0d140e9d2cefe57dc3a8e1508e2ab
 - any related extension versions

**Additional context**
The issue seems to be related to `lib/ash/plug_helpers.ex`:
```
if Code.ensure_loaded?(Plug.Conn) do
  defmodule Ash.PlugHelpers
```

If I remove the conditional, I get a warning from Ash saying that `Plug.Conn` is undefined:

```
==> ash
Compiling 322 files (.ex)
Compiling lib/ash/flow/dsl.ex (it's taking more than 10s)
warning: Plug.Conn.put_private/3 is undefined (module Plug.Conn is not available or is yet to be defined)
Invalid call found at 2 locations:
  lib/ash/plug_helpers.ex:44: Ash.PlugHelpers.set_actor/2
  lib/ash/plug_helpers.ex:105: Ash.PlugHelpers.set_tenant/2
```

Indeed, I can see that Plug is compiled *after* Ash.
But in this case AshHq compiles with only that warning.

Reading the documentation for `Code`: 
https://hexdocs.pm/elixir/1.14.0/Code.html#module-ensure_compiled-1-and-ensure_compiled-1

> Since Elixir's compilation happens in parallel, in some situations you may need to use a module that was not yet compiled, therefore it can't even be loaded.

It recommends to instead use `ensure_compiled!/1`:

> When invoked, [ensure_compiled/1](https://hexdocs.pm/elixir/1.14.0/Code.html#ensure_compiled/1) and [ensure_compiled!/1](https://hexdocs.pm/elixir/1.14.0/Code.html#ensure_compiled!/1) halt the compilation of the caller until the module becomes available. Note the distinction between [ensure_compiled/1](https://hexdocs.pm/elixir/1.14.0/Code.html#ensure_compiled/1) and [ensure_compiled!/1](https://hexdocs.pm/elixir/1.14.0/Code.html#ensure_compiled!/1) is important: if you are using [ensure_compiled!/1](https://hexdocs.pm/elixir/1.14.0/Code.html#ensure_compiled!/1), you are indicating to the compiler that you can only continue if said module is available.

However, it wouldn't resolve this bug:

> Finally, note you only need [ensure_compiled!/1](https://hexdocs.pm/elixir/1.14.0/Code.html#ensure_compiled!/1) to check for modules being defined within the same project. It does not apply to modules from dependencies as dependencies are always compiled upfront.

The only way to compilation order is with dependencies:
https://elixirforum.com/t/how-can-i-change-deps-compilation-order/24843

> You can not alter the order of compilation, it is determined automatically and make sure that dependencies are compiled before their dependents.

One solution might be to break out `Ash.PlugHelpers` into a separate library with Plug as a dependency,
And use PlugHelpers as a dependency for any library that needs access to those helpers.

**Comment by zachdaniel** [comment]
Great investigation! I think we may be able to fix it more simply with an optional dependency to `:plug` with `">= 0.0.0"`.

**Comment by zachdaniel** [comment]
This has been fixed in 424498ff8e7af1885258451064ab0b9b00e521f0
